### PR TITLE
Add download function

### PR DESCRIPTION
**Pull request by cromefire**
The PR is adding a download function that allows downloading of movies/... via android

### CODE REVIEW

**Comment by cromefire** [comment]
The android part _should_ be working, there's just some wire up into web required

**Comment by JustAMan** [comment]
What makes this a draft?

**Comment by dkanada** [comment]
@JustAMan it still needs to be hooked up to the web client.

**Comment by JustAMan** [comment]
That means no one can call this function right now, so what? To make a web hook you need something to call at. I think it should be marked as non-draft and merged before making a web hook.

**Comment by cromefire** [comment]
Well just add the hook to this PR you should have write permission

**Comment by nielsvanvelzen** [comment]
This PR will implement a [feature request](https://features.jellyfin.org/posts/323/download-option-for-android-app) that needs to be updated when the PR is merged / app update is released.

**Comment by cromefire** [comment]
but there might still be some hookup needed on this site, before it's mergable

**Comment by dkanada** [comment]
It does still need to be hooked up, but I attempted it already and apparently there's an issue with the dialog, even though the same dialog works fine for the battery prompt.

**Comment by cromefire** [comment]
What's the error?

File: src/NativeShell/src/NativeShell.java [hunk 1]
```diff
@@ -251,4 +254,56 @@ private boolean hideMediaSession() {
...
+                .setTitle(title)
+                .setDescription("Downloading...")
+                .setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, filename)
+                .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
+
+        cordova.getActivity().runOnUiThread(() -> {
+            // TODO translate these strings
+            new AlertDialog.Builder(context)
+                    .setTitle("Allowed network types")
+                    .setMessage("Do you want to allow the download to run over mobil data or roaming?\nCharges may apply!")
```
Comment (Derkades) [inline src/NativeShell/src/NativeShell.java]: Small typo "mobile"

**Comment by dkanada** [comment]
It's complaining that the alert can't be used unless a certain theme is used, but we can't control that with cordova so it would probably be easier to make a settings page and apply it there. I added a method to open a native settings page if you want to remove the dialog and move the network option to a settings activity.

**Comment by cromefire** [comment]
I'm currently on mobile, I'll look at it tomorrow or so

File: src/NativeShell/src/NativeShell.java [hunk 2]
```diff
@@ -251,4 +254,56 @@ private boolean hideMediaSession() {
...
+        Uri uri = Uri.parse(url);
+
+        DownloadManager.Request request = new DownloadManager.Request(uri)
+                .setTitle(title)
+                .setDescription("Downloading...")
+                .setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, filename)
+                .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
+
+        cordova.getActivity().runOnUiThread(() -> {
+            // TODO translate these strings
```
Comment (ferferga) [inline src/NativeShell/src/NativeShell.java:279]: When is this going to be done? IMHO, I wouldn't release any Android update without the strings being translatable. Can't we simply reuse the strings from Web?

Comment (cromefire) [reply src/NativeShell/src/NativeShell.java:279]: I don't know any of that... Depends on the architecture, I just copied the Todo from above

Comment (cromefire) [reply src/NativeShell/src/NativeShell.java:283]: There's no example how to use translated strings in this architecture, so I leave it to you guys

**Comment by cromefire** [comment]
It now ready from my side, it should work™

File: config.xml [hunk 3]
```diff
@@ -36,6 +36,5 @@
     </platform>
     <plugin name="cordova-plugin-whitelist" spec="^1.3.3" />
     <plugin name="cordova-plugin-chromecast" spec="https://github.com/jellyfin/cordova-plugin-chromecast" />
-    <plugin name="org.jellyfin.mobile" spec="src/NativeShell" />
```
Comment (JustAMan) [inline config.xml:39]: why?

File: src/NativeShell/src/NativeShell.java (hunk 2 above)
Comment (JustAMan) [reply src/NativeShell/src/NativeShell.java:283]: I would better have a non-translated feature that no feature at all, though.

File: config.xml (hunk 3 above)
Comment (cromefire) [reply config.xml:39]: Oh dev-link-setup thing

Comment (cromefire) [reply config.xml:39]: how can I turn that off?

Comment (JustAMan) [reply config.xml:39]: No idea, you can revert the change to that file for now though if it's not needed.

Comment (cromefire) [reply config.xml]: reverted it

Comment (cromefire) [reply config.xml]: it was part of that symlink to edit in android studio thing

**Comment by cromefire** [comment]
I've added some translations (en + de for now)

**Comment by cromefire** [comment]
Okay build is also working now... cordova is a crappy thing...

File: src/NativeShell/plugin.xml [hunk 4]
```diff
@@ -26,6 +26,24 @@
...
+    </config-file>
+    <config-file target="res/values-de/strings.xml" parent="/*">
+      <string name="battery_optimizations_title">Batterieoptimierungen deaktivieren</string>
+      <string name="battery_optimizations_message">Bitte die Batterieoptimierungen deaktivieren um Medien mit ausgeschaltetem Bildschirm abgespielt werden können.</string>
+      <string name="network_title">Erlaubte Verbindungsmöglichkeiten</string>
+      <string name="network_message">Wollen Sie den Download über Mobile Daten oder Roaming elauben?\\nEs können Kosten anfallen!</string>
+      <string name="wifi_only">Nur WLAN</string>
+      <string name="mobile_data">Mobile Daten</string>
+      <string name="mobile_data_and_roaming">Mobile Daten &amp; Roaming</string>
+    </config-file>
```
Comment (JustAMan) [inline src/NativeShell/plugin.xml:46]: is it possible to store those as separate files so we can push them to weblate?..

Comment (cromefire) [reply src/NativeShell/plugin.xml:46]: Doesn't seem so I've tried but it didn't build

Comment (cromefire) [reply src/NativeShell/plugin.xml:46]: Info on this is very spare...

**Comment by cromefire** [comment]
Does someone want to merge this?

File: src/NativeShell/src/NativeShell.java [hunk 5]
```diff
@@ -251,4 +255,59 @@ private boolean hideMediaSession() {
...
+        } catch (Exception e) {
+            callbackContext.error("error: download: " + e.getMessage());
+            return false;
+        }
+        Context context = cordova.getContext();
+        Uri uri = Uri.parse(url);
+
+        DownloadManager.Request request = new DownloadManager.Request(uri)
+                .setTitle(title)
+                .setDescription("Downloading...")
```
Comment (nielsvanvelzen) [inline src/NativeShell/src/NativeShell.java:277]: You're missing a string here

Comment (cromefire) [reply src/NativeShell/src/NativeShell.java]: Fixed

**Comment by dkanada** [comment]
@cromefire this will be in the next release for sure, I'm just working out the Azure issues first since that will block a release regardless.

**Comment by cromefire** [comment]
Yeah just keep in mind that it has to be included in web and has to be tested (which only works after it's in web)

**Comment by JustAMan** [comment]
I'm going to bite the bullet and merge it, so web part could be cooked.

**Comment by cromefire** [comment]
Should I open an issue there?

**Comment by JustAMan** [comment]
Yeah. Make one in jf-web and link to this PR pls.

**Pull request merged**